### PR TITLE
feat(gateway): add rewriteConnectionInitPayload option

### LINKED ISSUE #482: Allow rewriting of connectionInit payload for gateway
Hi,

similar to `rewriteHeaders`, it would be great if it would be possible to rewrite the `connection_init` payload before it is passed to the federated services. This way, all authentication can be handled within the gateway.

I would be happy to provide a PR, but would like your feedback on it first. My approach would be to provide an `rewriteConnectionInit` option, that accepts the `connection_init` payload received by the client and returns the one to be forwarded to the service. Another option would be to just extend the payload with the return of onConnect, but I think this would reduce transparency for the user.

Best
Philipp

### PR DESCRIPTION
**Pull request by steinroe**
closes #482

### CODE REVIEW

File: lib/subscription-client.js [hunk 1]
```diff
@@ -265,6 +267,11 @@ class SubscriptionClient {
       }
     }
 
+    let connectionInit = context?._connectionInit
```
[the remark this answers is not in the document]
Comment (steinroe) [reply lib/subscription-client.js]: Thanks for the hint! I removed the ?. and ?? operators and changed the type accordingly.

File: lib/subscription-client.js [hunk 2]
```diff
@@ -265,6 +267,14 @@ class SubscriptionClient {
       }
     }
 
+    let connectionInit
+    if (context) {
+      connectionInit = context._connectionInit
+    }
+    if (this.rewriteConnectionInitPayload) {
+      connectionInit = this.rewriteConnectionInitPayload(connectionInit, context)
```
[the remark this answers is not in the document]
Comment (PacoDu) [reply lib/subscription-client.js:275]: > I thought about it too, but wanted to adhere to the way it is handled with queries and mutations. For async operations, the user can use onConnect to fetch the data, add it to the context and read it from there in the rewrite method. Do you think it still makes sense to make the rewrite method async, too?

Good point ! The onConnect handler should be enough

**Review by mcollina — APPROVED** [review]
lgtm

**Comment by steinroe** [comment]
Can you publish a new release with this feature? :)

**Pull request merged**